server/rating_service/rating_service.py

```diff
@@ -0,0 +1,254 @@
+                    player_id,
+                    old_rating,
+                    new_rating,
+                )
+                await conn.execute(
+                    "UPDATE game_player_stats "
+                    "SET after_mean = %s, after_deviation = %s, scoreTime = NOW() "
+                    "WHERE gameId = %s AND playerId = %s",
+                    (new_rating.mu, new_rating.sigma, game_id, player_id),
+                )
+
+                await self._update_rating_table(
+                    conn, rating_type, player_id, new_rating, outcomes[player_id]
+                )
+
+                self._update_player_object(player_id, rating_type, new_rating)
+
+    def _update_player_object(
+        self, player_id: PlayerID, rating_type: RatingType, new_rating: Rating
+    ) -> None:
+        if self._player_service_callback is None:
+            self._logger.warning(
+                "Tried to send rating change to player service, "
+                "but no service was registered."
+            )
+            return
+
+        self._logger.debug(
+            "Sending player rating update for player with id ", player_id
+        )
+        self._player_service_callback(player_id, rating_type, new_rating)
+
+    async def _update_rating_table(
+        self,
+        conn,
+        rating_type: RatingType,
+        player_id: PlayerID,
+        new_rating: Rating,
+        outcome: GameOutcome,
+    ):
+        # If we are updating the ladder1v1_rating table then we also need to update
+        # the `winGames` column which doesn't exist on the global_rating table
+        table = f"{rating_type.value}_rating"
+
+        if rating_type is RatingType.LADDER_1V1:
+            is_victory = outcome is GameOutcome.VICTORY
+            await conn.execute(
+                "UPDATE ladder1v1_rating "
+                "SET mean = %s, is_active=1, deviation = %s, numGames = numGames + 1, winGames = winGames + %s "
+                "WHERE id = %s",
+                (new_rating.mu, new_rating.sigma, 1 if is_victory else 0, player_id),
+            )
+        else:
+            await conn.execute(
+                "UPDATE " + table + " "
+                "SET mean = %s, is_active=1, deviation = %s, numGames = numGames + 1 "
+                "WHERE id = %s",
+                (new_rating.mu, new_rating.sigma, player_id),
+            )
+
+    async def _join_rating_queue(self):
+        """
+        Offers a call that is blocking until the rating queue has been emptied.
+        Mostly for testing purposes.
+        """
+        await self._queue.join()
+
+    async def shutdown(self):
+        """
+        Finish rating all remaining games, then exit.
+        """
+        self._accept_input = False
+        self._logger.debug(
+            "Shutdown initiated. Waiting on current queue: %s", self._queue
+        )
+        await self._queue.join()
+        self._logger.debug("Queue emptied: %s", self._queue)
+        await self.kill()
```

Comment:
I wasn't sure if there's a benefit to calling `self._task.cancel()`.
I think I wanted to safeguard against having multiple worker threads in case you call `shutdown` and then `initialize` again, but actually I think that safeguarding should rather happen in `initialize`, I'll rework that...